node/src/main/scala/org/bitcoins/node/networking/peer/PeerMessageReceiver.scala

```diff
@@ -152,68 +163,73 @@ class PeerMessageReceiver(
     */
   private def handleControlPayload(
       payload: ControlPayload,
-      sender: PeerMessageSender): Try[Unit] = {
+      sender: PeerMessageSender): Future[PeerMessageReceiver] = {
     payload match {
 
       case versionMsg: VersionMessage =>
-        logger.trace(
-          s"Received versionMsg=${versionMsg}from peer=${peerOpt.get}")
+        logger(nodeAppConfig).trace(
+          s"Received versionMsg=${versionMsg}from peer=${peer}")
 
-        internalState match {
+        state match {
           case bad @ (_: Disconnected | _: Normal | Preconnection) =>
-            Failure(
+            Future.failed(
               new RuntimeException(
                 s"Cannot handle version message while in state=${bad}"))
 
           case good: Initializing =>
-            internalState = good.withVersionMsg(versionMsg)
+            val newState = good.withVersionMsg(versionMsg)
 
             sender.sendVerackMessage()
 
             //we want peers to just send us headers
             //we don't want to have to request them manually
             sender.sendHeadersMessage()
 
-            Success(())
+            val newRecv =
+              new PeerMessageReceiver(dataMessageHandler = dataMessageHandler,
+                                      state = newState,
+                                      peer = peer,
+                                      callbacks = callbacks)
```

Comment:
nit: This seems to be a regular pattern, maybe having a method on the old receiver that takes in a newState would be nice?